src/ai/defaultai.cc

```diff
@@ -4369,7 +4391,11 @@ bool DefaultAI::check_productionsites(uint32_t gametime) {
 
 	// No upgrade without proper workers
 	if (considering_upgrade && !site.site->has_workers(enhancement, game())) {
-		considering_upgrade = false;
+		const BuildingDescr& bld = *tribe_->get_building_descr(enhancement);
+		BuildingObserver& en_bo = get_building_observer(bld.name().c_str());
+			if (get_stocklevel(en_bo, gametime, WareWorker::kWorker) < 1) {
+				considering_upgrade = false;
```

Comment:
Well, dont we need to have "considering_update = true" somewhere?

Reply:
we have. see line 4325. Mechanism is that we declare that to true at the beginning and then we perform a lot of checks why it should not be done setting this to false. In this case I just added the check if a worker is available in warehouses. so this gets only false if no experienced/needed worker in warehouse and no suitable worker in site. The check of warehouses is necessary as experienced workers will be kicked out now.

Reply:
OK then...